### DISCUSSION TITLE
Where should I put utility files within my /api folder so that they don't count as functions

### DISCUSSION THREAD

**Discussion by TaylorFacen** [vercel/vercel#4983] [Help] [answered]
I have some files containing some helper code that's used in multiple api functions (e.g. db queries, common transformations, etc.). When I deploy my app, these files are counting as functions even though they're only in the api folder to assist my actual functions. How can I fix this so I don't go over my function limit. 

Thank you!

**Comment by markozxuu** [comment, marked as answer]
Hi @TaylorFacen 👋,  you could follow the following file structure:

root project:
 - api -> serverless function
 - utils ->  helper code

**Example** 

/api/index.js

`const sayHello = require(../utils/sayHello')
`

**Reply by styfle** [reply]
If you wish to have utils in the `/api` directory, you can prefix with an underscore or a dot such as `/api/_utils/foo.js`

**Reply by bokub** [reply]
I cannot find official documentation explaining this.

Would you have a link?

**Reply by TomasHubelbauer** [reply]
When I put a file in a directory with an underscore or prefix the file name itself with an underscore, then adjust the route to also use the underscore, I see Vercel still treats the file as an API route. Moving the file outside of the `api` directory makes it behave like a page when adjusting the route and navigating to it, so that's not a solution either. The only way I found that works is to place the helper files outside of `pages/api` completely.

**Reply by elitan** [reply]
> I cannot find official documentation explaining this.
> 
> Would you have a link?

https://vercel.com/docs/concepts/functions/serverless-functions#deploying-serverless-functions

**Reply by mauricewegner** [reply]
> https://vercel.com/docs/concepts/functions/serverless-functions#deploying-serverless-functions

The provided link does not contain any information about the underscore prefixing (anymore?). Is there any official information about it? I do remember it being documented a while ago.

**Reply by wiiiimm** [reply]
@mauricewegner _ prefixing is still officially supported and our documentation team will be adding this back to our docs soon. We will send you the URL here once it is ready.

**Reply by bokub** [reply]
@williamli 2 months later, and nothing in the documentation... Is underscore prefixing still supported or not? Thanks

**Reply by amyegan** [reply]
Files and directories that start with `_` are not turned into Serverless Functions, so you can prefix utility files in your `/api` directory with an underscore to avoid turning them into API endpoints.
https://github.com/vercel/vercel/blob/46c0fd153a23e7c4dc9a29d578f7683e42c10547/packages/fs-detectors/src/detect-builders.ts#L386-L388

~There is a PR open to add this mention back into the documentation and I expect it will be merged soon. Once the change is published, you'll be able to see it at https://vercel.com/docs/concepts/functions/serverless-functions#deploying-serverless-functions~

**Reply by molebox** [reply]
Hey folks, our docs have now been updated with this information! https://vercel.com/docs/concepts/functions/serverless-functions#adding-utility-files-to-the-/api-directory

**Comment by tomyo** [comment]
I want to elaborate in this question:
* Where do I place testing code for the api/ functions?

Usually code is for front end, except for code inside api/ directory, so, I would like to have tests for api/ living also inside api/, or

any recommended best practice? can't find this on docs.

the _ approach works for exporting/exporting inside the api/ functions, but not outside of them, as that results in calling the cloud function, not importing from the js file as desired.

Hope this makes sense, thanks

**Reply by amyegan** [reply]
Hi @tomyo! Where to put your testing code will depend on the library you're using and your configuration. There is an example of some integration and e2e test coverage of a Next.js app with Playwright at https://github.com/vercel/examples/tree/main/solutions/testing

You might also want to check out some examples that people have shared elsewhere online, such as [Paige Niedringhaus' detailed walkthrough](https://www.paigeniedringhaus.com/blog/how-to-unit-test-next-js-api-routes-with-typescript) and [Sean Connolly's concise overview](https://seanconnolly.dev/unit-testing-nextjs-api-routes)

**Reply by tomyo** [reply]
Thanks a lot @amyegan!
:)

**Comment by danrasmuson** [comment]
I've discovered that `import` from `/api/_utils/foo.js` does not work out of the box with create-react-app

My guess into why this is the case is because of the difference between...

**non-create-react-app `.vercel/output/builds.json`** 

```
{
  "//": "This file was generated by the `vercel build` command. It is not part of the Build Output API.",
  "target": "preview",
  "argv": [
    "/Users/danielrasmuson/.nvm/versions/node/v16.18.0/bin/node",
    "/Users/danielrasmuson/.nvm/versions/node/v16.18.0/bin/vercel",
    "build"
  ],
  "builds": [
    {
      "require": "@vercel/node",
      "requirePath": "/Users/danielrasmuson/.nvm/versions/node/v16.18.0/lib/node_modules/vercel/node_modules/@vercel/node/dist/index",
      "apiVersion": 3,
      "use": "@vercel/node",
      "src": "api/test.ts",
      "config": {
        "zeroConfig": true
      }
    },
    {
      "require": "@vercel/static",
      "requirePath": "",
      "apiVersion": 2,
      "use": "@vercel/static",
      "src": "!{api/**,package.json,middleware.[jt]s}",
      "config": {
        "zeroConfig": true
      }
    }
  ]
}
```

**create-react-app `.vercel/output/builds.json`** 

```
{
  "//": "This file was generated by the `vercel build` command. It is not part of the Build Output API.",
  "target": "preview",
  "argv": [
    "/Users/danielrasmuson/.nvm/versions/node/v16.18.0/bin/node",
    "/Users/danielrasmuson/.nvm/versions/node/v16.18.0/bin/vercel",
    "build"
  ],
  "builds": [
    {
      "require": "@vercel/node",
      "requirePath": "/Users/danielrasmuson/.nvm/versions/node/v16.18.0/lib/node_modules/vercel/node_modules/@vercel/node/dist/index",
      "apiVersion": 3,
      "use": "@vercel/node",
      "src": "api/test.ts",
      "config": {
        "zeroConfig": true
      }
    },
    {
      "require": "@vercel/static-build",
      "requirePath": "/Users/danielrasmuson/.nvm/versions/node/v16.18.0/lib/node_modules/vercel/node_modules/@vercel/static-build/dist/index",
      "apiVersion": 2,
      "src": "package.json",
      "use": "@vercel/static-build",
      "config": {
        "zeroConfig": true,
        "framework": "create-react-app"
      }
    }
  ]
}
```

The different between the two build outputs is `@vercel/static` and `@vercel/static-build`

The `@vercel/static` excludes everything in the `api/**` folder

```
"src": "!{api/**,package.json,middleware.[jt]s}",
```

While `@vercel/static-build` has no such condition



I'm not yet aware of how to get `import` in `/api` to function properly in the create-react-app build context.
